test/tests_for_sdkmr/EE2Runjob_test.py

```diff
@@ -0,0 +1,268 @@
+    # added.
+    if not hasattr(got_job, "id"):
+        assert not hasattr(expected_job, "id")
+    else:
+        assert got_job.id == expected_job.id
+    assert got_job.user == expected_job.user
+    assert got_job.authstrat == expected_job.authstrat
+    assert got_job.wsid == expected_job.wsid
+    assert got_job.status == expected_job.status
+
+    # The Job class fills the updated field with the output of time.time on instantiation
+    # so we can't do a straight equality
+    assert abs(got_job.updated - expected_job.updated) < 1
+
+    assert got_job.queued == expected_job.queued
+    assert got_job.estimating == expected_job.estimating
+    assert got_job.running == expected_job.running
+    assert got_job.finished == expected_job.finished
+    assert got_job.errormsg == expected_job.errormsg
+    assert got_job.msg == expected_job.msg
+    assert got_job.error == expected_job.error
+    assert got_job.terminated_code == expected_job.terminated_code
+    assert got_job.error_code == expected_job.error_code
+    assert got_job.scheduler_type == expected_job.scheduler_type
+    assert got_job.scheduler_id == expected_job.scheduler_id
+    assert got_job.scheduler_estimator_id == expected_job.scheduler_estimator_id
+
+    if not got_job.job_input:
+        assert expected_job.job_input is None
+    else:
+        assert got_job.job_input.wsid == expected_job.job_input.wsid
+        assert got_job.job_input.method == expected_job.job_input.method
+        assert (
+            got_job.job_input.requested_release
+            == expected_job.job_input.requested_release
+        )
+        assert got_job.job_input.params == expected_job.job_input.params
+        assert got_job.job_input.service_ver == expected_job.job_input.service_ver
+        assert got_job.job_input.app_id == expected_job.job_input.app_id
+        assert (
+            got_job.job_input.source_ws_objects
+            == expected_job.job_input.source_ws_objects
+        )
+        assert got_job.job_input.parent_job_id == expected_job.job_input.parent_job_id
+
+        assert got_job.job_input.requirements.clientgroup == (
+            expected_job.job_input.requirements.clientgroup
+        )
+        assert (
+            got_job.job_input.requirements.cpu
+            == expected_job.job_input.requirements.cpu
+        )
+        assert (
+            got_job.job_input.requirements.memory
+            == expected_job.job_input.requirements.memory
+        )
+        assert (
+            got_job.job_input.requirements.disk
+            == expected_job.job_input.requirements.disk
+        )
+        assert got_job.job_input.requirements.estimate == (
+            expected_job.job_input.requirements.estimate
+        )
+
+        # this fails, which should be impossible given all the fields above pass
+        # assert got_job.job_input.requirements == expected_job.job_input.requirements
```

Comment:
are there other fields around that you're not accounting for?

Reply:
I don't think so: https://github.com/kbase/execution_engine2/blob/develop/lib/execution_engine2/db/models/models.py#L151-L165

Reply:
Which test fails when calling this?

Reply:
Did you try any of the `unittest` equality functions (e.g. https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertEqual) in case they give better diagnostics of the problem?

Reply:
> Which test fails when calling this?
The commented out assert fails

> Have you tried any of the unittest equality functions (e.g. https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertEqual) in case they give better diagnostics of the problem?

Since I'm not subclassing `TestCase`, which contains the `assertEquals()` method, I don't think I can, although if you know how to handle that without having to restructure the tests into a class I'll give it a go. In any case, in my experience, `pytest` with `assert`s give much better diagnostics than `unittest`.

Reply:
Sure, I see the commented out assert, but you call the `assert_jobs_equal` function twice - line 120 and line 155. Do they both always fail? Or just one or the other? It could have to do with how you're mocking. The `expected_job` has its `narrative_cell_info` set to a `Meta()` object, while `got_job` might have it as an `EmbeddedDocumentField` wrapped around a `Meta()`. If they don't have an `__eq__` function that does duck-typing against its values, but also takes in the class, it might get wiggy when you compare the outer classes like that. I don't know enough about MongoEngine to know what's happening.

Also one or the other class might have extra attributes set.

Reply:
> you call the assert_jobs_equal function twice - line 120 and line 155. Do they both always fail?

Neither fail, both pass.

Reply:
> The expected_job has its narrative_cell_info set to a Meta() object, while got_job might have it as an EmbeddedDocumentField wrapped around a Meta().

If that were true, comparing the fields of the Meta() object to the fields of the EDF wrapper should cause the test to fail.

Reply:
>  I don't know enough about MongoEngine to know what's happening

Reply:
> Also one or the other class might have extra attributes set.

Reply:
here's some more fun:

```
            print(vars(got_job.job_input.requirements))
            print(got_job.job_input.requirements.estimate)
            print(repr(got_job.job_input.requirements.estimate))
            print(type(got_job.job_input.requirements.estimate))
            print(vars(expected_job.job_input.requirements))
            print(expected_job.job_input.requirements.estimate)
            print(repr(expected_job.job_input.requirements.estimate))
            print(type(expected_job.job_input.requirements.estimate))
    
            # this fails, which should be impossible given all the fields above pass
>           assert got_job.job_input.requirements == expected_job.job_input.requirements
```

Results in:

```
{'_cls': 'JobRequirements'}
None
None
<class 'NoneType'>
{'_cls': 'JobRequirements'}
None
None
<class 'NoneType'>
```

So unless there's some invisible field set by the superclass I'm stumped.

Reply:
There are only the 5 fields in JobRequirements: https://github.com/kbase/execution_engine2/blob/develop/lib/execution_engine2/db/models/models.py#L139-L148

Reply:
Neither have an `id` field

Reply:
I got the embedded documents to all pass normally - it was a `lib` in an import line again.

I still can't get the outermost `Job` class to pass an equality check even when the fields are identical.

Reply:
Ok, so far I've
- confirmed equals for the the `job_input` field works - this is the only field that is or contains embedded objects
- set the job_input field in `got_job` and `expected_job` to `None`
- confirmed the tests still fail when trying to use a direct assert even when the fields appear identical:
```
>       assert got_job == expected_job
E       assert {"user": "someuser", "authstrat": "kbaseworkspace", "status": "created", "updated": 1614120410.3491368, "child_jobs": [], "batch_job": false} == {"user": "someuser", "authstrat": "kbaseworkspace", "status": "created", "updated": 1614120410.3491368, "child_jobs": [], "batch_job": false}
E         +{"user": "someuser", "authstrat": "kbaseworkspace", "status": "created", "updated": 1614120410.3491368, "child_jobs": [], "batch_job": false}
E         -{"user": "someuser", "authstrat": "kbaseworkspace", "status": "created", "updated": 1614120410.3491368, "child_jobs": [], "batch_job": false}
```
- confirmed the types and import paths for the Job objects are the same
- confirmed the types and values of all the fields in the field list are the same

At this point I'm out of ideas.

Reply:
Well, here's the problem: 
https://github.com/MongoEngine/mongoengine/blob/master/mongoengine/base/document.py#L284-L297

Reply:
The hierarchy is `Job` -> `Document` -> `BaseDocument`. Neither `Job` nor `Document` have an `__eq__` method. In contrast, `JobInput` and the other embedded classes inherit from `EmbeddedDocument`: https://github.com/MongoEngine/mongoengine/blob/master/mongoengine/document.py#L85-L88

Reply:
Oh and all these classes use `__slots__` hence no dict